funding/batch.go

```diff
@@ -0,0 +1,512 @@
+package funding
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/psbt"
+	"github.com/lightningnetwork/lnd/labels"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/walletrpc"
+	"github.com/lightningnetwork/lnd/lnwallet/chanfunding"
+)
+
+var (
+	// errShuttingDown is the error that is returned if a signal on the
+	// quit channel is received which means the whole server is shutting
+	// down.
+	errShuttingDown = errors.New("shutting down")
+)
+
+// batchChannel is a struct that keeps track of a single channel's state within
+// the batch funding process.
+type batchChannel struct {
+	fundingReq    *InitFundingMsg
+	pendingChanID [32]byte
+	updateChan    chan *lnrpc.OpenStatusUpdate
+	errChan       chan error
+	fundingAddr   string
+	chanPoint     *wire.OutPoint
+	isPending     bool
+}
+
+// processPsbtUpdate processes the first channel update message that is sent
+// once the initial part of the negotiation has completed and the funding output
+// (and therefore address) is known.
+func (c *batchChannel) processPsbtUpdate(u *lnrpc.OpenStatusUpdate) error {
+	psbtUpdate := u.GetPsbtFund()
+	if psbtUpdate == nil {
+		return fmt.Errorf("got unexpected channel update %v", u.Update)
+	}
+
+	if psbtUpdate.FundingAmount != int64(c.fundingReq.LocalFundingAmt) {
+		return fmt.Errorf("got unexpected funding amount %d, wanted "+
+			"%d", psbtUpdate.FundingAmount,
+			c.fundingReq.LocalFundingAmt)
+	}
+
+	c.fundingAddr = psbtUpdate.FundingAddress
+
+	return nil
+}
+
+// processPendingUpdate is the second channel update message that is sent once
+// the negotiation with the peer has completed and the channel is now pending.
+func (c *batchChannel) processPendingUpdate(u *lnrpc.OpenStatusUpdate) error {
+	pendingUpd := u.GetChanPending()
+	if pendingUpd == nil {
+		return fmt.Errorf("got unexpected channel update %v", u.Update)
+	}
+
+	hash, err := chainhash.NewHash(pendingUpd.Txid)
+	if err != nil {
+		return fmt.Errorf("could not parse outpoint TX hash: %v", err)
+	}
+
+	c.chanPoint = &wire.OutPoint{
+		Index: pendingUpd.OutputIndex,
+		Hash:  *hash,
+	}
+	c.isPending = true
+
+	return nil
+}
+
+// RequestParser is a function that parses an incoming RPC request into the
+// internal funding initialization message.
+type RequestParser func(*lnrpc.OpenChannelRequest) (*InitFundingMsg, error)
```

Comment:
Nice series of type defs here along with the interfaces to make this easier to test 👍